src/openms/source/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.cpp

```diff
@@ -0,0 +1,314 @@
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Ahmed Khalil $
+// $Authors: Ahmed Khalil $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.h>
+
+namespace OpenMS
+{
+  IsotopeLabelingMDVs::IsotopeLabelingMDVs() :
+    DefaultParamHandler("IsotopeLabelingMDVs")
+  {
+  }
+
+  IsotopeLabelingMDVs::~IsotopeLabelingMDVs()
+  {
+  }
+
+  void IsotopeLabelingMDVs::updateMembers_()
+  {
+  }
+  
+  void IsotopeLabelingMDVs::isotopicCorrection(
+    const Feature& normalized_feature,
+    Feature& corrected_feature,
+    const std::vector<std::vector<double>> correction_matrix,
+    const std::string correction_matrix_agent)
+  {
+    // MDV_corrected = correction_matrix_inversed * MDV_observed (normalized_features)
+    
+    std::vector<std::vector<double>> selected_correction_matrix;
+    auto correction_matrix_search = correction_matrices_.find(correction_matrix_agent);
+    
+    if (!correction_matrix_agent.empty() && correction_matrix_search != correction_matrices_.end())
+    {
+      selected_correction_matrix = correction_matrix_search->second ;
+    }
+    else
+    {
+      selected_correction_matrix = correction_matrix;
+    }
+    
+    uint16_t correction_matrix_n = selected_correction_matrix.size();
+    std::vector<std::vector<double>> correction_matrix_inversed(correction_matrix_n, std::vector<double>(correction_matrix_n,0));
+    
+    // 1- correction matrix inversion
+    Eigen::MatrixXd correction_matrix_eigen(selected_correction_matrix.size(), selected_correction_matrix[0].size());
+    for (uint i = 0; i < selected_correction_matrix.size(); i++){
+      for (uint j = 0; j < selected_correction_matrix[0].size(); j++){
+        correction_matrix_eigen(i,j) = selected_correction_matrix[i][j];
+      }
+    }
+    Eigen::MatrixXd correction_matrix_eigen_inversed = correction_matrix_eigen.inverse();
+    for (int i = 0; i < correction_matrix_eigen_inversed.rows(); i++){
+      for (int j = 0; j < correction_matrix_eigen_inversed.cols(); j++){
+        correction_matrix_inversed[i][j] = correction_matrix_eigen_inversed(i,j);
```

Comment:
e.g., is this copying needed? calculations below could also be done directly on eigen matrix

Reply:
unnecessary copying has been removed